lib/fw.c

```diff
@@ -1260,16 +1262,35 @@ static int switchtec_fw_img_write_hdr_gen4(int fd,
  * @param[in]  fd	File descriptor for image file to write
  * @param[in]  info	Partition information structure
  * @return 0 on success, error code on failure
+ *
+ * The offset of image body in the image file is greater than or equal to
+ * the image header length. This function also repositions the read/write
+ * file offset of fd to the offset of image body in the image file if
+ * needed. This will facilitate the switchtec_fw_read_fd() function which
+ * is usually called following this function to complete a firmware image
+ * read.
  */
 int switchtec_fw_img_write_hdr(int fd, struct switchtec_fw_image_info *info)
 {
+	int ret;
+
 	switch (info->gen) {
-	case SWITCHTEC_GEN3: return switchtec_fw_img_write_hdr_gen3(fd, info);
-	case SWITCHTEC_GEN4: return switchtec_fw_img_write_hdr_gen4(fd, info);
+	case SWITCHTEC_GEN3:
+		ret = switchtec_fw_img_write_hdr_gen3(fd, info);
+		break;
+	case SWITCHTEC_GEN4:
+		ret = switchtec_fw_img_write_hdr_gen4(fd, info);
+		lseek(fd, info->part_body_offset, SEEK_SET);
```

Comment:
Why not just do this inside of switchtec_fw_img_write_hdr_gen4()? Looks like it would be a lot cleaner.

Reply:
I thought it would make the code more clear that we see hdr handling difference of gen3 and gen4 at this level. On the other hand, yes it would be cleaner to get it into write_hdr_gen4(). Will update.